File: java/src/yarin/yal/graph/RootedTree.java
package yarin.yal.graph;

import java.util.ArrayList;
import java.util.Arrays;

public class RootedTree {
    private final Node root;
    private final Node[] rootedTreeNodes;
    private final Node[] topologicalOrder; // root = topologicalOrder[0]
    private Node[][] lcaPreprocess;

    public RootedTree(Graph.Node root) {
        if (root == null) {
            throw new IllegalArgumentException();
        }
        Graph graph = root.getGraph();
        if (graph.getEdges().size() != graph.getNodes().size() - 1) {
            throw new IllegalArgumentException();
        }
        int n = graph.getNodes().size();
        rootedTreeNodes = new Node[n];

        // Build up tree structure without using recursion and create a topological ordering
        int[] level = new int[n];
        int[] order = new int[n];
        int[] parent = new int[n];
        level[root.getIndex()] = 0;
        order[0] = root.getIndex();
        parent[root.getIndex()] = -1;
        for (int i = 0, j = 1; i < n; i++) {
            if (i == j) {
                throw new IllegalArgumentException();
            }
            for (Graph.Node node : graph.getNode(order[i]).getAdjacentNodes()) {
                int nodeIndex = node.getIndex();
                if (nodeIndex != root.getIndex() && level[nodeIndex] == 0) {
                    parent[nodeIndex] = order[i];
                    level[nodeIndex] = level[order[i]] + 1;
                    order[j++] = nodeIndex;
                }
            }
        }

        topologicalOrder = new Node[n];
        for (int i = n - 1; i >= 0; i--) {
            int nodeIndex = order[i];
            topologicalOrder[i] = new Node(graph.getNode(nodeIndex), parent[nodeIndex], level[nodeIndex]);
        }

        this.root = topologicalOrder[0];
    }

    public Node getRoot() {
        return this.root;
    }

    public Node getNode(int index) {
        return rootedTreeNodes[index];
    }

    private void setNode(Node treeNode) {
        rootedTreeNodes[treeNode.getNode().getIndex()] = treeNode;
    }

    private Node[][] preprocessLCA() {
        int N = rootedTreeNodes.length, M = 0;
        while ((1 << M) < N) {
            M++;
        }
        Node[][] p = new Node[N][M];

        for (int i = 0; i < N; i++) {
            p[i][0] = rootedTreeNodes[i].getParent();
        }

        for (int j = 1; 1 << j < N; j++) {
            for (int i = 0; i < N; i++) {
                if (p[i][j - 1] != null) {
                    p[i][j] = p[p[i][j - 1].getNode().getIndex()][j - 1];
                }
            }
        }

        return p;
    }

    public Node getLowestCommonAncestor(Node p, Node q) {
        if (lcaPreprocess == null) {
            lcaPreprocess = preprocessLCA();
        }

        if (p.getLevel() < q.getLevel()) {
            Node tmp = p;
            p = q;
            q = tmp;
        }

        int log;
        for (log = 1; 1 << log <= p.getLevel(); log++) ;
        log--;

        for (int i = log; i >= 0; i--) {
            if (p.getLevel() - (1 << i) >= q.getLevel()) {
                p = lcaPreprocess[p.getNode().getIndex()][i];
            }
        }

        if (p == q) {
            return p;
        }

        for (int i = log; i >= 0; i--) {
            if (lcaPreprocess[p.getNode().getIndex()][i] != null &&
                lcaPreprocess[p.getNode().getIndex()][i] != lcaPreprocess[q.getNode().getIndex()][i])
            {
                p = lcaPreprocess[p.getNode().getIndex()][i];
                q = lcaPreprocess[q.getNode().getIndex()][i];
            }
        }

        return p.getParent();
    }

    public class Node {
        private final Graph.Node node;
        private final Node[] children;
        private final int parentIndex;
        private final int level;

        public Graph.Node getNode() {
            return this.node;
        }

        public Node getParent() {
            return parentIndex < 0 ? null : RootedTree.this.getNode(parentIndex);
        }

        public int getLevel() {
            return this.level;
        }

        public Node[] getChildren() {
            return this.children;
        }

        public Node(Graph.Node node, int parentIndex, int level) {
            this.node = node;
            this.level = level;
            this.parentIndex = parentIndex;

            ArrayList<Node> children = new ArrayList<Node>(node.getDegree());
            for (Graph.Node child : node.getAdjacentNodes()) {
                if (child.getIndex() != parentIndex) {
                    children.add(RootedTree.this.getNode(child.getIndex()));
                }
            }
            this.children = children.toArray(new Node[0]);

            RootedTree.this.setNode(this);
        }

        @Override
        public String toString() {
            int[] children = new int[this.children.length];
            for (int i = 0; i < this.children.length; i++) {
                children[i] = this.children[i].getNode().getIndex();
            }
            return String.format("[index=%s; parent=%d; children={%s}]",
                                 node.getIndex(), parentIndex, Arrays.toString(children));
        }
    }
}

File: cs/Graphs/RootedTree.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Algorithms.Graphs
{
	public class RootedTree
	{
		public RootedTreeNode Root { get; private set; }
		private readonly RootedTreeNode[] RootedTreeNodes;
		private RootedTreeNode[,] lcaPreprocess;
		private readonly RootedTreeNode[] topologicalOrder; // root = topologicalOrder[0]

		public RootedTree(Graph graph, Node root)
		{
			if (graph == null || root == null || root.Graph != graph || graph.Edges.Count != graph.Nodes.Count - 1)
				throw new ArgumentException();
			var n = graph.Nodes.Count;
			RootedTreeNodes = new RootedTreeNode[n];

			// Build up tree structure without using recursion and create a topological ordering
			var level = new int[n];
			var order = new int[n];
			var parent = new int[n];
			level[root.Index] = 0;
			order[0] = root.Index;
			parent[root.Index] = -1;
			for (int i = 0, j = 1; i < n; i++)
			{
				if (i == j)
					throw new ArgumentException();
				foreach (var node in graph.GetNode(order[i]).GetAdjacentNodes())
				{
					var nodeIndex = node.Index;
					if (nodeIndex != root.Index && level[nodeIndex] == 0)
					{
						parent[nodeIndex] = order[i];
						level[nodeIndex] = level[order[i]] + 1;
						order[j++] = nodeIndex;
					}
				}
			}

			topologicalOrder = new RootedTreeNode[n];
			for (int i = n - 1; i >= 0; i--)
			{
				var nodeIndex = order[i];
				topologicalOrder[i] = new RootedTreeNode(this, graph.GetNode(nodeIndex), parent[nodeIndex], level[nodeIndex]);
			}

			Root = topologicalOrder[0];
		}

		public RootedTreeNode GetNode(int index)
		{
			return RootedTreeNodes[index];
		}

		internal void SetNode(RootedTreeNode treeNode)
		{
			RootedTreeNodes[treeNode.Node.Index] = treeNode;
		}

		private RootedTreeNode[,] PreprocessLCA()
		{
			int N = RootedTreeNodes.Length, M = 0;
			while ((1 << M) < N)
				M++;
			var p = new RootedTreeNode[N, M];

			for (int i = 0; i < N; i++)
				p[i, 0] = RootedTreeNodes[i].Parent;

			for (int j = 1; 1 << j < N; j++)
				for (int i = 0; i < N; i++)
					if (p[i, j - 1] != null)
						p[i, j] = p[p[i, j - 1].Node.Index, j - 1];

			return p;
		}

		public RootedTreeNode GetLowestCommonAncestor(RootedTreeNode p, RootedTreeNode q)
		{
			if (lcaPreprocess == null)
				lcaPreprocess = PreprocessLCA();

			if (p.Level < q.Level)
			{
				RootedTreeNode tmp = p;
				p = q;
				q = tmp;
			}

			int log;
			for (log = 1; 1 << log <= p.Level; log++) ;
			log--;

			for (int i = log; i >= 0; i--)
				if (p.Level - (1 << i) >= q.Level)
					p = lcaPreprocess[p.Node.Index, i];

			if (p == q)
				return p;

			for (int i = log; i >= 0; i--)
				if (lcaPreprocess[p.Node.Index, i] != null &&
					lcaPreprocess[p.Node.Index, i] != lcaPreprocess[q.Node.Index, i])
				{
					p = lcaPreprocess[p.Node.Index, i];
					q = lcaPreprocess[q.Node.Index, i];
				}

			return p.Parent;
		}
	}

	public class RootedTreeNode
	{
		private readonly int _parentIndex;

		public RootedTree Tree { get; private set; }
		public Node Node { get; private set; }
		public RootedTreeNode Parent { get { return _parentIndex < 0 ? null : Tree.GetNode(_parentIndex); } }
		public int Level { get; private set; }

		public RootedTreeNode[] Children { get; private set; }

		public RootedTreeNode(RootedTree tree, Node node, int parentIndex, int level)
		{
			Tree = tree;
			Node = node;
			Level = level;
			_parentIndex = parentIndex;

			var children = new List<RootedTreeNode>(node.Degree);
			foreach (Node child in Node.GetAdjacentNodes().Where(x => x.Index != parentIndex))
				children.Add(tree.GetNode(child.Index));
			Children = children.ToArray();

			tree.SetNode(this);
		}
	}
}
